Self-Consistent-Field/RHF_EFP.py

```diff
@@ -0,0 +1,376 @@
+                  'scf_type': 'pk',
+                  'e_convergence': 1e-8})
+
+# Set defaults
+maxiter = 40
+E_conv = 1.0E-6
+D_conv = 1.0E-3
+
+# Integral generation from Psi4's MintsHelper
+wfn = psi4.core.Wavefunction.build(mol, psi4.core.get_global_option('BASIS'))
+t = time.time()
+mints = psi4.core.MintsHelper(wfn.basisset())
+S = np.asarray(mints.ao_overlap())
+
+# Get nbf and ndocc for closed shell molecules
+nbf = S.shape[0]
+ndocc = wfn.nalpha()
+
+print('\nNumber of occupied orbitals: %d' % ndocc)
+print('Number of basis functions: %d' % nbf)
+
+# Run a quick check to make sure everything will fit into memory
+I_Size = (nbf ** 4) * 8.e-9
+print("\nSize of the ERI tensor will be %4.2f GB." % I_Size)
+
+# Estimate memory usage
+memory_footprint = I_Size * 1.5
+if I_Size > numpy_memory:
+    psi4.core.clean()
+    raise Exception("Estimated memory utilization (%4.2f GB) exceeds numpy_memory \
+                    limit of %4.2f GB." % (memory_footprint, numpy_memory))
+
+# Compute required quantities for SCF
+V = np.asarray(mints.ao_potential())
+T = np.asarray(mints.ao_kinetic())
+I = np.asarray(mints.ao_eri())
+
+print('\nTotal time taken for integrals: %.3f seconds.' % (time.time() - t))
+
+t = time.time()
+
+# Build H_core
+H = T + V
+
+# <-- efp: add in permanent moment contribution and cache
+Vefp = modify_Fock_permanent(mol, nbf, efpmol)
+assert(psi4.compare_integers(1, np.allclose(Vefp, ref_V2), 'EFP permanent Fock contrib'))
+H = H + Vefp
+Horig = H.copy()
+set_qm_atoms(mol, efpmol)
+# --> efp
+
+# Orthogonalizer A = S^(-1/2) using Psi4's matrix power.
+A = mints.ao_overlap()
+A.power(-0.5, 1.e-16)
+A = np.asarray(A)
+
+# Calculate initial core guess
+Hp = A.dot(H).dot(A)
+e, C2 = np.linalg.eigh(Hp)
+C = A.dot(C2)
+Cocc = C[:, :ndocc]
+D = np.einsum('pi,qi->pq', Cocc, Cocc)
+
+print('\nTotal time taken for setup: %.3f seconds' % (time.time() - t))
+
+print('QM/EFP: iterating Total Energy including QM/EFP Induction')
+t = time.time()
+E = 0.0
+Enuc = mol.nuclear_repulsion_energy()
+Eold = 0.0
+Dold = np.zeros_like(D)
+
+for SCF_ITER in range(1, maxiter + 1):
+
+    # <-- efp: add contribution to Fock matrix
+    H = Horig
+    verbose_dipoles = 1 if (SCF_ITER == 1) else 0
+    # [Kaliman:2013:2284] Fig. 4 -- Compute electric field from wavefunction
+    # [Kaliman:2013:2284] Fig. 4 -- Compute electric field from induced dipoles
+    Vefp = modify_Fock_induced(nbf, efpmol, verbose=verbose_dipoles)
+    H = H + Vefp
+    # --> efp
+
+    # Build fock matrix
+    J = np.einsum('pqrs,rs->pq', I, D)
+    K = np.einsum('prqs,rs->pq', I, D)
+    F = H + J * 2 - K
+
+    diis_e = np.einsum('ij,jk,kl->il', F, D, S) - np.einsum('ij,jk,kl->il', S, D, F)
+    diis_e = A.dot(diis_e).dot(A)
+
+    # SCF energy and update
+    # [Kaliman:2013:2284] Fig. 4 -- Compute QM wavefunction
+    SCF_E = np.einsum('pq,pq->', F + H, D) + Enuc
+    dRMS = np.mean(diis_e**2)**0.5
+
+    # <-- efp: add contribution to energy
+    density = D
```

Comment:
What does this do here?

Reply:
It's a hack. The libefp callback requires a specific signature. Can't get the density into it through args, so it has to be named what the fn insides expect to access as "global". I'll explain it.

Reply:
Ah, I see. Maybe call it `efp_density` and inside that functional do a `global efp_density` to help make people aware?

Reply:
never had to use the dreaded `global` before. sure.